### ISSUE TITLE
Two-way autosync of any folder

### ISSUE THREAD

**Issue by kepi** [haiwen/seadroid#55]
It will be huge benefit of Seafile android app if it can sync selected folders similar as it does on desktop.

Same future can be used to auto upload of camera photos but also for many other use cases (saved notes, scanned documents etc). But not only upload but syncing, so I just take 50 photos of same thing because I can't decide :) and I can simply delete those 49 unfocused at my computer and have them deleted also on phone.

As for UI, user should be able to just select any folder in Seafile and than select sync with folder on phone. Than he simply choose any folder on his android and syncing will start.

Setting for enabling only on Wifi should be present.

You can see same/similar functionality for dropbox in external app:
https://play.google.com/store/apps/details?id=com.ttxapps.dropsync

**Comment by dszymczuk** [comment]
:+1:

**Comment by schokonsti** [comment]
+1 for two-way-sync.
With this functionality the android client would be more powerful than the desktop :+1:

**Comment by mrjovanovic** [comment]
Does this function not exist already? Is the Seadroid app just an app for browsing files on the server and downloading them? What happens to individual files opened/downloaded in the app?

I am seriously asking, because the documentation doesn't seem to cover this use case (an incredibly useful one).

**Comment by Logan676** [comment]
> Does this function not exist already?
To be simple, Seadroid App (latest version 1.5.1) doesn\`t support two-way auto sync of any folders yet. 

> Is the Seadroid app just an app for browsing files on the server and downloading them?
Not really, Seadroid App is more powerful than that, e.g. camera auto upload

> What happens to individual files opened/downloaded in the app?
If a file was downloaded, it will be added to a background observer queue, any modifications on the file will be auto uploaded (one way sync) to the server.

**Comment by mrjovanovic** [comment]
@Logan676 
Thanks so much for the replies. Clears something up. :) I agree, sync in the same way as it exists in the desktop client would be extremely useful.

How would I go about adding this information to the [wiki](https://seacloud.cc/group/3/wiki/)?

**Comment by moritzschaefer** [comment]
What's the current state of this? Can I help in some way to make this feature happen?

**Comment by Logan676** [comment]
@moritzschaefer 
Currently, it supports one way sync of a single file, any modifications on the file will be auto uploaded to the cloud.
Now we plan to support one way sync of any folder, any modifications under the folder will be auto uploaded.

It\`s more than welcome if you can help us make it happen, no matter by submitting patches or functional prototypes or anything else relevant.

Steps to make this happen may looks like this
1. add a "sync" menu to drop down menu of each folder
2. monitor modifications specifically under the selected folder
3. auto sync when detecting changes

Please let me know if you need any help.

**Comment by forouher** [comment]
I wonder if it might be possible to implement the full seafile HTTP sync protocol in Seadroid. That would solve many problems, including #301, #277 and #91.

- One would have to keep the ".seafile-data" folder as small as possible. But at a first glance that might be doable.
- Always syncing a full repository would not be an option, obviously. One would need to figure out a way to do downloads/merge/upload/... individual files or directories without having to checkout the whole repository. I don't know whether that's doable. Maybe using hidden sub-libraries?
- There would have to be a clever sync scheduler, that only syncs files when necessary. (User opens a file, user refreshes a dirent, wifi+ a/c power available,...)

Maybe there are more roadblocks that I've missed. However if these turn out to be doable, using JNI it might even be possible to re-use some of the C code from the desktop client.

**Comment by jult** [comment]
For crying out loud, this is still not in there? Geez, that's very disappointing. Should be the first feature it should have. Why don't you use code from syncthing, they've done it all in their first android release, and it's all on github.

**Comment by raffh** [comment]
What is the status of this issue?
It would be a great feature to have and making seadroid 100% usable.

**Comment by serathius** [comment]
+1 for 2 way sync

**Comment by infinity85** [comment]
Would like this feature soooooo much, as well!

At least perhaps something like one-way-but-the-other-way-around... hence you could backup your whatsapp folder (always on internal Android memory in "**Whatsapp**" directory. For this example it is not even necessary to have it two way, but just like photoupload... simply upload everything that is in this folder every day/week/month... just to replace the google drive backup feature, which is not encrypting anything I guess.

**Comment by LeonardoGentile** [comment]
any update on this?

**Comment by LeonardoGentile** [comment]
ping

**Comment by ikcalB** [comment]
+1

**Comment by infinity85** [comment]
+1 too

**Comment by christianfl** [comment]
+1 This feature would be awesome! I'd like to synchronize txt-notes for a very long time..

**Comment by infinity85** [comment]
Even a manual "sync" button for a certain directory would be at least A solution... But the demand is simply ignored :(

**Comment by f0086** [comment]
+1

**Comment by Ralosso** [comment]
+3

maybe for some people  there is another way. 
If u enable the seafdav server, you can use apps like foldersync to sync in both ways. But this is very slow and sometimes a bit buggy.

**Comment by christianfl** [comment]
But FolderSync is no free software. So this is no alternative for me.

**Comment by ZEALi** [comment]
+1 for two-way-sync like what pc client does.

**Comment by Raphi111** [comment]
+1 I also need to synchronize .txt to be able to drop Google Drive

**Comment by droidzone** [comment]
+1 for this

**Comment by SkaveRat** [comment]
sooo.... nothing? I just installed seafile and I'm really surprised that this is the intended behaviour

**Comment by f0086** [comment]
What exactly is the problem not implementing this core feature?
Is it just that there is no time for that or is the solution too complex to tackle?

**Comment by End-to-End-is-the-way** [comment]
I would love this feature... Any updates?

**Comment by sesom42** [comment]
It is a pity that nothing has been going on for three years. Seafile advertises with "enterprise file sync" but this function is not available in seadroid (except for photos, contacts and individual files). For a simple download you could use the website. So why use an app? Two-way sync is an essential feature for a sync app.

**Comment by rabyte** [comment]
I aggree, I would love this feature. Especially since it works for photos already I naively thought it should not be too tricky.
Maybe there are some road blockers I am overlooking...?

**Comment by Cisco30** [comment]
+1 for this

**Comment by ydylla** [comment]
+1
This would make the seafile ecosystem perfect

**Comment by mxinden** [comment]
+1

**Comment by infinity85** [comment]
+1!

**Comment by henrythemouse** [comment]
WOW, no comment from a Seafile rep for over 4 years! My mother used to tell me "If you can't say anything good then don't say anything at all." Maybe that's the issue.
+1

**Comment by hydrandt** [comment]
+1 for this feature
looking to start using https://github.com/laurent22/joplin/ for notes, and hoped to use seafile (which I use for photos upload from mobile) for the desktop/mobile sync. Now it seems I'll have to use syncthing to do that.
Maybe we could do https://www.bountysource.com/ , if resources are the problem?

**Comment by gshegosh** [comment]
After spending 2 hours on installing SeaFile in docker on my server and then wrestling with inability to log in from Android client (turns out my password was... too long), now I find that there's no way to sync files which breaks my intended use case of having my Keepass file everywhere. Disappointing :(

**Comment by henrythemouse** [comment]
As an alternative method you can use a webdav client (such as FolderSync). Seafile server can be configured to accept webdav connections and  FolderSync can sync any files/folders you have on your android device.

**Comment by forouher** [comment]
@gshegosh In case you're planning to use Keepass2Droid - that app actually works very well together with Seadroid.

Starting with Android 4.4 there exists another method by which Android apps can access files stored in Seafile, its called the "Storage Access Framework".

In Keepass2Droid, select "Open file", and then "Android File browser". There, select you're Seafile server on the left. Then select your keepass file. That's it. You'll have a full two-way-sync. :)

**Comment by End-to-End-is-the-way** [comment]
I fully agree with @gshegosh. A full sync with Android with the Seafile Android client is long overdue. There is a way, however, to fully sync file and folders on an Android phone via Google Drive (and other cloud services). I've written a description here: https://jochen-plikat.com/2017/12/29/cloud-speicher-mit-android-tablet-synchronisieren/ (in German). But I would certainly prefer to do this with the Seafile Client and end-to-end-encryption. It's very disappointing that there is no comment whatsoever by the Seafile developers.

**Comment by gshegosh** [comment]
Thank you for your suggestions. I did think about WebDav and configured it yesterday. Keepass2Droid opens it directly, without the need for FolderSync. Google Drive - I don't trust them and there's no sync client on Linux (there's Insync, but it's a third party option, additionally paid).

Forouher's suggestion to use Storage Access Framework is nice, I'm using it now. It doesn't seem to sync back from the phone to the server, but I don't really need to add passwords from the phone. Thanks! It allows me to disable webdav access, and it's one thing less to secure.

**Comment by End-to-End-is-the-way** [comment]
I agree with you, gshegosh, I wouldn't use Google Drive to Sync the KeePass Database (or any other sensitive data) either. It is, however, useful for a full sync of less sensitive data if you don't use a Linux desktop. I use Google Drive to sync articles (two-way, full sync) in pdf format between my tablet and my Mac. But, as I said, I hope this will one day be possible with end-to-end-encryption (be it Seafile or any other Open Source app).

**Comment by wepanx** [comment]
Is this feature planned for being integrated some day / this year?

There are indeed enterprises which (also) use android devices and not only desktop PCs/Macs..... @
"Seafile - **Enterprise file sync** and share platform with high reliability and performance" ;-)

**Comment by puck** [comment]
> In Keepass2Droid, select "Open file", and then "Android File browser". There, select your Seafile server on the left. Then select your keepass file. That's it. You'll have a full two-way-sync. :)

Oh, this worked nicely in OI Safe to backup my password file. However, it doesn't seem to work for over writing an existing file. This is Android 9, running the Seafile android app v2.2.8.

**Comment by abh153k** [comment]
It's such a shame that the android client doesn't have an option to auto-upload files from __any__ directory. Had the client had this simple utility, I won't have to be look for other alternatives.

**Comment by bugf1x** [comment]
any update on this? Why can we use two-way sync on any platform except android? It works using 3rd party apps like synchronize ultimate.. but why not with the native client? This is a huge flaw.

**Comment by chekoopa** [comment]
Still relevant.

**Comment by chb567** [comment]
I would love to see this feature, im currently using syncthing for this. 
Syncthing has the options available on android as on desktop exposed through the standard 'web' ui. I wish to leverage the full potential of seafile realized on android in the future.

**Comment by pmdevita** [comment]
Any chance this is being worked on or that someone could lay out what needs to be done to make this happen?
